lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+            if lock is not None:
+                err = "{0} exception when removing failure mark for {1}: {2}"
+                msg = 'Removing failure mark on {0}'
+                try:
+                    tty.verbose(msg.format(pkg_id))
+                    lock.release_write()
+                except AssertionError:
+                    pass
+                except Exception as exc:
+                    tty.warn(err.format(exc.__class__.__name__, pkg_id,
+                                        str(exc)))
+
+    def _cleanup_task(self, pkg, remove_task):
+        """
+        Cleanup the build task for the spec
+
+        Args:
+            pkg (spack.package.PackageBase): the package being installed
+            remove_task (bool): ``True`` if the build task should be removed
+        """
+        if remove_task:
+            self._remove_task(pkg.unique_id)
+
+        # Ensure we have a read lock to prevent others from uninstalling the
+        # spec during our installation.
+        self._ensure_read_locked(pkg)
+
+    def _ensure_install_ready(self, pkg):
+        """
+        Ensure the package is ready to install locally, which includes
+        already locked.
+
+        Args:
+            pkg (spack.package.PackageBase): the package being locally
+                installed
+        """
+        pre = "{0} cannot be installed locally:".format(pkg.unique_id)
+
+        # External packages cannot be installed locally.
+        if pkg.spec.external:
+            # TODO: If keep, use a custom exception
+            raise spack.error.SpackError('{0} {1}'.format(pre, 'is external'))
+
+        # Upstream packages cannot be installed locally.
+        if pkg.installed_upstream:
+            # TODO: If keep, use a custom exception
+            raise spack.error.SpackError('{0} {1}'.format(pre, 'is upstream'))
+
+        # The package must have a prefix lock at this stage.
+        if pkg.unique_id not in self.locks:
+            # TODO: If keep, use a custom exception
+            raise spack.error.SpackError('{0} {1}'.format(pre, 'not locked'))
+
+    def _ensure_read_locked(self, pkg):
+        """
+        Add a prefix read lock for the package spec, downgrading from write if
+        needed.
+
+        Args:
+            pkg (spack.package.PackageBase): the package whose spec is being
+                installed
+        """
+        pkg_id = pkg.unique_id
+        ltype, lock = self.locks[pkg_id] if pkg_id in self.locks \
+            else ('read', None)
+        nolock = lock is None
+        if nolock or ltype == 'write':
+            msg = '{0} a read lock on {1} with timeout {2}'
+            err = 'Failed to {0} a read lock for {1} due to {2}: {3}'
+
+            # Wait until the other process finishes if there are no more
+            # build tasks with priority 0 (i.e., with no uninstalled
+            # dependencies).
+            no_p0 = len(self.build_tasks) <= 0 or not self._next_is_pri0()
```

Comment:
Why not:
```suggestion
            no_p0 = len(self.build_tasks) == 0 or not self._next_is_pri0()
```
?

Reply:
Habit.  Had a professor who pushed covering cases in conditionals.  (I know length will/should never be < 0.)